### PR TITLE
ARCO-170: replace go-bt with go-sdk

### PR DESCRIPTION
**Pull request by pawellewandowski98**
# Description

- replaced go-bt with go-sdk
- go-bt cannot be removed entirely because go-bc (which cannot be replaced - not every method is in go-sdk) is using it

### CODE REVIEW

File: internal/beef/beef_test.go [hunk 1]
```diff
@@ -235,7 +227,6 @@ func TestDecodeBEEF_InvalidBeef_ReturnError(t *testing.T) {
...
...0503bb7a8704fdb6295b3001b5e8876a101fd3d0300aeea966733175ff60b55bc77edcb83c0fce3453329f51195e5cbc7a874ee47ad01fd9f0100f67f50b53d73ffd6e84c02ee1903074b9a5b2ac64c508f7f26349b73cca9d7e901ce006ce74c7beed0c61c50dda8b578f0c0dc5a393e1f8758af2fb65edf483afcaa68016600e32475e17bdd141d62524d0005989dd1db6ca92c6af70791b0e4802be4c5c8c1013200b88162f494f26cc3a1a4a7dcf2829a295064e93b3dbb2f72e21a73522869277a011800a938d3f80dd25b6a3a80e450403bf7d62a1068e2e4b13f0656c83f764c55bb77010d006feac6e4fea41c37c508b5bfdc00d582f6e462e6754b338c95b448df37bd342c010700bf5448356be23b2b9afe53d00cee047065bbc16d0bbcc5f80aa8c1b509e45678010200c2e37431a437ee311a737aecd3caae1213db353847f33792fd539e380bdb4d440100005d5aef298770e2702448af2ce014f8bfcded5896df5006a44b5f1b6020007aeb01010091484f513003fcdb25f336b9b56dafcb05fbc739593ab573a2c6516b344ca5320101000000018ae36502fdc82837319362c488fb9cb978e064daf600bbfc48389663fc5c160c000000006b48304502210095ea412e82881f81de63764f38f2562bd8a184b06686b3a9e4a5a8b47b9ea1cf022018c76a08b46168c8beb3f7e508cd21c208882b1ec801a9ad7c1b791995092429412102b0c8980f5d2cab77c92c68ac46442feba163a9d306913f6a34911fc618c3c4e7ffffffff01f4010000000000001976a9148a8c4546a95e6fc8d18076a9980d59fd882b4e6988ac0000000000"
```
Comment (kuba-4chain) [inline internal/beef/beef_test.go:238]: Why did we remove this case?

File: internal/fees/fees.go [hunk 2]
```diff
@@ -1,23 +1,61 @@
...
+// Change Description:
+// Previously, the fee calculation used `math.Ceil` to round up the transaction size
+// to the nearest kilobyte, which resulted in charging the full fee for a kilobyte
+// even for transactions smaller than 1000 bytes. For example, a transaction with a
+// size of 500 bytes would be charged the same fee as a 1000-byte transaction.
+//
+// The updated implementation removes the rounding, so the fee is now directly
+// proportional to the transaction size. For instance, a transaction with 500 bytes
+// will now be charged 50% of the fee for a full kilobyte, ensuring a fairer and
+// more accurate fee calculation.
```
Comment (kuba-4chain) [inline internal/fees/fees.go:32]: Nice!

File: internal/validator/default/default_validator.go [hunk 3]
```diff
@@ -114,19 +114,20 @@ func cumulativeCheckFees(ctx context.Context, txFinder validator.TxFinderI, tx *
...
+		cumulativeSize += tx.Size()
+		//cumulativeSize.TotalBytes += size.TotalBytes
+		//cumulativeSize.TotalDataBytes += size.TotalDataBytes
+		//cumulativeSize.TotalStdBytes += size.TotalStdBytes
 
 		cumulativePaidFee += tx.TotalInputSatoshis() - tx.TotalOutputSatoshis()
 	}
 
-	expectedFee, err := validator.CalculateMiningFeesRequired(&cumulativeSize, feeQuote)
+	//expectedFee, err := validator.CalculateMiningFeesRequired(&cumulativeSize, feeQuote)
```
Comment (kuba-4chain) [inline internal/validator/default/default_validator.go:129]: I think we can remove this comment

File: internal/validator/default/default_validator.go [hunk 4]
```diff
@@ -139,8 +140,9 @@ func cumulativeCheckFees(ctx context.Context, txFinder validator.TxFinderI, tx *
 	return nil
 }
 
-func isFeePaidEnough(fees *bt.FeeQuote, tx *bt.Tx) (bool, uint64, uint64, error) {
-	expFeesPaid, err := validator.CalculateMiningFeesRequired(tx.SizeWithTypes(), fees)
+func isFeePaidEnough(feeModel transaction.FeeModel, tx *transaction.Transaction) (bool, uint64, uint64, error) {
+	//expFeesPaid, err := validator.CalculateMiningFeesRequired(tx.SizeWithTypes(), fees)
```
Comment (kuba-4chain) [inline internal/validator/default/default_validator.go:144]: Remove this comment

File: internal/beef/beef_test.go [hunk 5]
```diff
@@ -3,11 +3,10 @@ package beef
 import (
 	"encoding/hex"
 	"errors"
+	"github.com/bitcoin-sv/go-sdk/script"
+	"github.com/bitcoin-sv/go-sdk/transaction"
```
Comment (boecklim) [inline internal/beef/beef_test.go:7]: nit: move these imports below next to the other 3rd party dependencies

File: internal/beef/beef_test.go (hunk 1 above)
Comment (pawellewandowski98) [reply internal/beef/beef_test.go:238]: go-sdk do not have this check

File: internal/beef/bump.go [hunk 6]
```diff
@@ -2,40 +2,26 @@ package beef
...
-			}
-			if blockMerkleRoot == "" {
-				blockMerkleRoot = mr
-			} else if blockMerkleRoot != mr {
-				return nil, errors.New("different merkle roots for the same block")
-			}
-		}
-
-		if blockMerkleRoot == "" {
-			return nil, errors.New("no transactions marked as expected to verify in bump")
```
Comment (boecklim) [inline internal/beef/bump.go:33]: Why is this not needed anymore? Is it already handled in `go-sdk`?

Comment (boecklim) [reply internal/beef/bump.go:33]: Ah I see it is in a separate function 👍

File: internal/beef/bump_test.go [hunk 7]
```diff
@@ -2,9 +2,9 @@ package beef
 
 import (
 	"errors"
+	"github.com/bitcoin-sv/go-sdk/transaction"
```
Comment (boecklim) [inline internal/beef/bump_test.go:5]: nit: move to other 3rd party imports

File: internal/beef/beef_test.go [hunk 8]
```diff
@@ -210,7 +202,7 @@ func TestDecodeBEEF_DecodeBEEF_HandlingErrors(t *testing.T) {
 		},
 		{
 			name:                "unable to decode BUMP leaf - not enough bytes for hash - proper BEEF marker, number of bumps, block height and tree height, nLeaves, offset and flag but with not enough bytes for hash",
-			hexStream:           "0100beef01fe8a6a0c000c04fde80b0011774f01d26412f0d16ea3f0447be0b5ebec67b0782e321a7a01cbdf7f734e",
+			hexStream:           "0100beef01fe8a6a0c000c04fde80b0011774f01d26412f0d16ea3f0447be0b5ebec67b0782e321a7a0",
```
[the remark this answers is not in the document]
Comment (pawellewandowski98) [reply internal/beef/beef_test.go:205]: Returned to old version but go-sdk do not have one check which go-bc had:

```
if len(bytes) < skip+32 {
    return nil, 0, errors.New("BUMP bytes do not contain enough data to be valid")
}
```

Without this check it tries to read data that is not there and gives another error, I have added it and replaced it in the test

File: internal/beef/beef.go [hunk 9]
```diff
@@ -83,31 +82,37 @@ func DecodeBEEF(beefHex []byte) (*BEEF, []byte, error) {
...
 		if err != nil {
 			return nil, nil, err
 		}
 
-		beefBytes = beefBytes[bytesUsed:]
+		// calculate the number of bytes used to encode the bump
+		usedBytes := beefBytes[:len(bump.Bytes())]
+		if !bytes.Equal(bump.Bytes(), usedBytes) {
+			return nil, nil, errors.New("error")
+		}
```
[the remark this answers is not in the document]
Comment (pawellewandowski98) [reply internal/beef/beef.go]: Done, also updated error message, I left this one after simple tests and forgot to change it

File: internal/validator/default/default_validator.go [hunk 10]
```diff
@@ -114,19 +114,20 @@ func cumulativeCheckFees(ctx context.Context, txFinder validator.TxFinderI, tx *
...
 
 	for _, tx := range txSet {
-		size := tx.SizeWithTypes()
-		cumulativeSize.TotalBytes += size.TotalBytes
-		cumulativeSize.TotalDataBytes += size.TotalDataBytes
-		cumulativeSize.TotalStdBytes += size.TotalStdBytes
+		cumulativeSize += tx.Size()
+		//cumulativeSize.TotalBytes += size.TotalBytes
+		//cumulativeSize.TotalDataBytes += size.TotalDataBytes
+		//cumulativeSize.TotalStdBytes += size.TotalStdBytes
```
[the remark this answers is not in the document]
Comment (pawellewandowski98) [reply internal/validator/default/default_validator.go]: Removed

File: internal/validator/default/default_validator.go (hunk 3 above)
Comment (pawellewandowski98) [reply internal/validator/default/default_validator.go]: Removed

File: internal/validator/default/default_validator.go (hunk 4 above)
Comment (pawellewandowski98) [reply internal/validator/default/default_validator.go]: Removed

File: internal/broadcaster/rate_broadcaster.go [hunk 11]
```diff
@@ -186,15 +166,17 @@ utxoLoop:
...
+			err = PayTo(tx, b.ks.Script, amount)
 			if err != nil {
 				return nil, fmt.Errorf("failed to pay transaction %d: %v", amount, err)
 			}
 
-			// Todo: Add OP_RETURN with text "ARC testing" so that WoC can tag it
+			err = tx.AddOpReturnOutput([]byte("ARC testing"))
+			if err != nil {
+				return nil, err
+			}
```
Comment (boecklim) [inline internal/broadcaster/rate_broadcaster.go:177]: Let's do this separately. There is a task for it
The text should be variable and it should be tested with WoC if it works as intended.

File: internal/validator/default/default_validator_test.go [hunk 12]
```diff
@@ -117,11 +116,13 @@ func TestValidator(t *testing.T) {
...
-			parentTx, err := bt.NewTxFromString(parentHex)
+			parentTx, err := transaction.NewTransactionFromHex(parentHex)
 			require.NoError(t, err, "Could not parse parent tx hex")
 
-			in.PreviousTxSatoshis = parentTx.Outputs[in.PreviousTxOutIndex].Satoshis
-			in.PreviousTxScript = parentTx.Outputs[in.PreviousTxOutIndex].LockingScript
+			in.SetPrevTxFromOutput(parentTx.Outputs[in.SourceTxOutIndex])
+
+			//in.PreviousTxSatoshis = parentTx.Outputs[in.PreviousTxOutIndex].Satoshis
+			//in.PreviousTxScript = parentTx.Outputs[in.PreviousTxOutIndex].LockingScript
```
Comment (boecklim) [inline internal/validator/default/default_validator_test.go:125]: nit: remove comments

File: internal/beef/beef_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (kuba-4chain) [reply internal/beef/beef_test.go]: Nice, thanks!

**Review by kuba-4chain — APPROVED** [review]

File: internal/validator/default/default_validator_test.go (hunk 12 above)
Comment (pawellewandowski98) [reply internal/validator/default/default_validator_test.go]: Removed

File: internal/broadcaster/rate_broadcaster.go (hunk 11 above)
Comment (pawellewandowski98) [reply internal/broadcaster/rate_broadcaster.go]: Sure, I came back to TODO comment

File: internal/beef/bump_test.go (hunk 7 above)
Comment (pawellewandowski98) [reply internal/beef/bump_test.go]: Done

File: internal/beef/beef_test.go (hunk 5 above)
Comment (pawellewandowski98) [reply internal/beef/beef_test.go]: Done, refactored all imports by gci linter

**Review by boecklim — APPROVED** [review]

**Pull request merged**